codex/stores/repostore.nim

```diff
@@ -41,119 +45,277 @@ declareGauge(codex_repostore_bytes_reserved, "codex repostore bytes reserved")
-    cid = ? Cid.init(data[sizeof(uint64)..<sizeof(uint64) + n]).mapFailure
-  success(cid)
+proc encode(t: Cid): seq[byte] = t.data.buffer
+proc decode(T: type Cid, bytes: seq[byte]): ?!Cid = Cid.init(bytes).mapFailure
 
-method putBlockCidAndProof*(
-  self: RepoStore,
-  treeCid: Cid,
-  index: Natural,
-  blockCid: Cid,
-  proof: MerkleProof
-): Future[?!void] {.async.} =
-  ## Put a block to the blockstore
-  ##
+proc encode(t: QuotaUsage): seq[byte] = t.autoencode
+proc decode(T: type QuotaUsage, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: BlockMetadata): seq[byte] = t.autoencode
+proc decode(T: type BlockMetadata, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: LeafMetadata): seq[byte] = t.autoencode
+proc decode(T: type LeafMetadata, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+###########################################################
+# Helper types and procs
+###########################################################
+
+type
+  DeleteResultKind = enum
+    Deleted = 0,    # block removed from store
+    InUse = 1,      # block not removed, refCount > 0 and not expired
+    NotFound = 2    # block not found in store
+
+  DeleteResult = object
+    kind: DeleteResultKind
+    released: NBytes
+
+  StoreResultKind = enum
+    Stored = 0,         # new block stored
+    AlreadyInStore = 1  # block already in store
+
+  StoreResult = object
+    kind: StoreResultKind
+    used: NBytes
+
+proc encode(t: DeleteResult): seq[byte] = t.autoencode
+proc decode(T: type DeleteResult, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: StoreResult): seq[byte] = t.autoencode
+proc decode(T: type StoreResult, bytes: seq[byte]): ?!T = T.autodecode(bytes)
 
+proc putLeafMetadata(self: RepoStore, treeCid: Cid, index: Natural, blkCid: Cid, proof: MerkleProof): Future[?!StoreResultKind] {.async.} =
   without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
     return failure(err)
 
-  let value = (blockCid, proof).encode()
+  await self.metaDs.modifyTGetU(key,
+    proc (maybeCurrMd: ?LeafMetadata): Future[(?LeafMetadata, StoreResultKind)] {.async.} =
+      var
+        md: LeafMetadata
+        res: StoreResultKind
 
-  await self.metaDs.put(key, value)
+      if currMd =? maybeCurrMd:
+        md = currMd
+        res = AlreadyInStore
+      else:
+        md = LeafMetadata(blkCid: blkCid, proof: proof)
+        res = Stored
 
-proc getCidAndProof(
-  self: RepoStore,
-  treeCid: Cid,
-  index: Natural
-): Future[?!(Cid, MerkleProof)] {.async.} =
+      (md.some, res)
+  )
+
+proc getLeafMetadata(self: RepoStore, treeCid: Cid, index: Natural): Future[?!LeafMetadata] {.async.} =
   without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
     return failure(err)
 
-  without value =? await self.metaDs.get(key), err:
+  without leafMd =? await self.metaDs.get(key), err:
     if err of DatastoreKeyNotFound:
       return failure(newException(BlockNotFoundError, err.msg))
```

Comment:
I thought about it, thing is it wouldn't make anything simpler. 

Ideally would be to use `Option[T]` but then the return type would be `Future[Result[Option[T], E]]` or using questionable it would be `Future[?![?T]]`. In my opinion that's just too long and too complicated. 
Currently we always return `failure(BlockNotFoundError)` for both blocks and blocks metadata and since it always translates to the same behaviour - handling missing block routine - we don't need to introduce another type for `failure(BlockMetadataNotFoundError)`, making the handling uniform in the client code (as it would be with handling `T.none` if we would use `Option[T]`).